core/sdk-core/src/test/java/software/amazon/awssdk/core/endpointdiscovery/EndpointDiscoveryRefreshCacheTest.java

```diff
@@ -0,0 +1,91 @@
+
+    private EndpointDiscoveryRefreshCache endpointDiscoveryRefreshCache;
+    private EndpointDiscoveryCacheLoader mockClient;
+    private static final URI testURI = URI.create("test_endpoint");
+    private static final String requestCacheKey = "request_cache_key";
+    private static final String accessKey = "access_cache_key";
+
+    @BeforeEach
+    public void setup() {
+         this.mockClient= mock(EndpointDiscoveryCacheLoader.class);
+         this.endpointDiscoveryRefreshCache = EndpointDiscoveryRefreshCache.create(mockClient);
+    }
+
+    @Test
+    public void getAsync_notRequired_returns_CompletedFuture() throws ExecutionException, InterruptedException {
+        when(mockClient.discoverEndpoint(any())).thenReturn(new CompletableFuture<>());
+        EndpointDiscoveryRequest request = EndpointDiscoveryRequest.builder()
+                                                                   .required(false)
+                                                                   .defaultEndpoint(testURI)
+                                                                   .build();
+        assertThat(endpointDiscoveryRefreshCache.getAsync("key", request).isDone()).isEqualTo(true);
+        assertThat(endpointDiscoveryRefreshCache.getAsync("key", request).get()).isEqualTo(testURI);
+
+    }
+
+    @Test
+    public void getAsync_returns_CompletedFuture() throws ExecutionException, InterruptedException {
+
+        when(mockClient.discoverEndpoint(any())).thenReturn(new CompletableFuture<>());
+        EndpointDiscoveryRequest request = EndpointDiscoveryRequest.builder()
+                                                                   .required(true)
+                                                                   .defaultEndpoint(testURI)
+                                                                   .build();
+        CompletableFuture<URI> future = endpointDiscoveryRefreshCache.getAsync("key", request);
+        assertThat(future.isDone()).isEqualTo(false);
+
+        future.complete(testURI);
+
+        assertThat(future.isDone()).isEqualTo(true);
+        assertThat(future.get()).isEqualTo(testURI);
+    }
+
+    @Test
+    public void getAsync_future_cancelled() {
+
+        when(mockClient.discoverEndpoint(any())).thenReturn(new CompletableFuture<>());
+        EndpointDiscoveryRequest request = EndpointDiscoveryRequest.builder()
+                                                                   .required(true)
+                                                                   .defaultEndpoint(testURI)
+                                                                   .build();
+        CompletableFuture<URI> future = endpointDiscoveryRefreshCache.getAsync("key", request);
+        assertThat(future.isDone()).isEqualTo(false);
+
+        future.cancel(true);
+        assertThat(future.isCancelled()).isEqualTo(true);
+        assertThrows(CancellationException.class, () -> future.get());
```

Comment:
nit: prefer to use `assertThatThrownBy` since it's also part of assertj which is used for everything else.